executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::HashSet;
+
+use powdr_ast::{analyzed::AlgebraicReference, asm_analysis::Machine};
+use powdr_number::FieldElement;
+
+use crate::witgen::{machines::MachineParts, FixedData};
+
+use super::{
+    affine_symbolic_expression::Effect,
+    cell::Cell,
+    witgen_inference::{ReferenceEvaluator, WitgenInference},
+};
+
+struct BlockMachineProcessor<'a, T: FieldElement> {
+    fixed_data: &'a FixedData<'a, T>,
+    machine_parts: MachineParts<'a, T>,
+    block_size: usize,
+}
+
+impl<T: FieldElement> BlockMachineProcessor<'_, T> {
+    pub fn generate_code(&self) -> Vec<Effect<T, Cell>> {
+        // TODO
+        let known_cells = Vec::new();
+        let ref_eval = ReferenceEvaluatorForFixedData(self.fixed_data);
+        let mut witgen = WitgenInference::new(self.fixed_data, ref_eval, known_cells);
+        let mut complete = HashSet::new();
+        while complete.len() != self.machine_parts.identities.len() * self.block_size {
+            for row in 0..self.block_size {
+                for id in self.machine_parts.identities.iter() {
+                    if !complete.contains(&(id.id(), row))
+                        && witgen.process_identity(id, row as i32)
+                    {
+                        complete.insert((id.id(), row));
+                    }
+                }
+            }
+        }
+        witgen.code()
+    }
+}
+
+// TODO: Copied from witgen_inference.rs, probably FixedData should just implement ReferenceEvaluator.
```

Comment:
I don't think this would be a good idea. We need two different evaluators depending on the situation:
1) for a block machine, depending on how we translate absolute row indices to relative row indices (I found it convenient to have offset '0' be the latch row, but maybe not), the translation has to be done here. We could also check that we only evaluate fixed columns that are cyclic, but we should probably do that when we construct the block machine.
2) for a dynamic machine, the first row needs to use absolute row indices into the fixed columns (if we ever want to fully abandon the interpreter, we need to implement that case as well)
3) for a dynamic machine for the "update" function, we need to check that we access fixed columns only in places where they are constant.

Points 2 and 3 mainly apply to the `is first` and `is last` columns.